### ISSUE TITLE
Computed properties do not type check when they could

### ISSUE THREAD

**Issue by leoasis** [Microsoft/TypeScript#18596]
**TypeScript Version:**  2.4.2

**Code**

```ts
type A = { a: number, b: string };

const a: A = { a: 1, b: 'foo' };

const b: A = { ...a, b: 123 }; // doesn't type check, b should be string (correct)

const propB: 'b' = 'b';
const c: A = { ...a, [propB]: 123 }; // type checks even though i'm setting b to a number (incorrect)
```

**Expected behavior:**

Last statement should be invalid, since `propB` is of type `'b'`, which should be enough information for TS to know that i'm setting the `b` property.

**Actual behavior:**

Last statement is valid for TS. This also happens for union types of string or number literals. When the types for the values in those keys are the same, and attempting to set a value using a computed property, TS allows invalid values.

**Comment by andy-ms** [comment]
Looks like this was fixed by #18317. In `typescript@next` your last line is an error.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
